src/Rings/mpoly-graded.jl

```diff
@@ -1612,6 +1612,349 @@ function Base.show(io::IO, h::HilbertData)
+  for e in still_kept
+    v = copy(e)
+    v[j] = (e[j] > k ? e[j] - k : 0)
+    push!(result, v)
+  end
+  return result
+end
+
+function _find_maximum(a::Vector{Int})
+  m = a[1]
+  j = 1
+  for i in 1:length(a)
+    if a[i] > m 
+      m = a[i]
+      j = i
+    end
+  end
+  return j
+end
+
+function _cmp_revlex(a::Vector{Int}, b::Vector{Int})
+  for i in 1:length(a)
+    a[i] == b[i] && continue
+    return a[i] > b[i]
+  end
+  return true
+end
+
+function _sort_lex(a::Vector{Vector{Int}})
+  pivot = a[1]
+  less = Vector{Vector{Int}}()
+  more = Vector{Vector{Int}}()
+  for m in a[2:end]
+    if _cmp_revlex(pivot, m)
+      push!(less, m)
+    else
+      push!(more, m)
+    end
+  end
+  return vcat(less, [pivot], more)
+end
+
+function _hilbert_numerator_from_leading_exponents(
+    a::Vector{Vector{Int}};
+    weight_matrix= [1 for i in 1:1, j in 1:length(a[1])],
+    return_ring=(all(x->(x>=0), weight_matrix) ? 
+                 PolynomialRing(QQ, [Symbol("t_$i") for i in 1:nrows(weight_matrix)])[1] : 
+                 LaurentPolynomialRing(QQ, [Symbol("t_$i") for i in 1:nrows(weight_matrix)])[1]
+                ),
+    #alg=:generator
+    #alg=:custom
+    #alg=:gcd
+    #alg=:indeterminate
+    #alg=:cocoa
+    alg=:BayerStillmanA # This is by far the fastest strategy. Should be used.
+  )
+  length(a) == 0 && return one(return_ring)
+
+  t = gens(return_ring)
+
+  # See Proposition 5.3.6
+  _are_pairwise_coprime(a) && return prod([1-prod([t[i]^(sum([e[j]*weight_matrix[i, j] for j in 1:length(e)])) for i in 1:length(t)]) for e in a])
+
+  if alg == :BayerStillmanA
+    S = return_ring
+    ###########################################################################
+    # For this strategy see
+    #
+    # Bayer, Stillman: Computation of Hilber Series
+    # J. Symbolic Computation (1992) No. 14, pp. 31--50
+    #
+    # Algorithm 2.6, page 35
+    ###########################################################################
+    r = length(a)
+
+    # End of recursion
+    r == 0 && return one(S)
+
+    # make sure we have lexicographically ordered monomials
+    a = _sort_lex(a)
+
+
+    # initialize the result 
+    h = one(S) - prod([t[i]^sum([a[1][j]*weight_matrix[i, j] for j in 1:length(a[1])]) for i in 1:length(t)])
+    for i in 2:r
+      J = _divide_by(a[1:i-1], a[i])
+      J1 = Vector{Vector{Int}}()
+      linear_mons = Vector{Int}()
+      for m in J
+        if sum(m) > 1
+          push!(J1, m)
+        else
+          for k in 1:length(m)
+            if m[k] > 0 
+              push!(linear_mons, k)
+              break
+            end
+          end
+        end
+      end
+      q = _hilbert_numerator_from_leading_exponents(J1, return_ring=S, 
+                                                    weight_matrix=weight_matrix, 
+                                                    alg=alg)
+      for k in linear_mons
+        q = q*(one(S) - prod([t[i]^weight_matrix[i, k] for i in 1:length(t)]))
```

Comment:
The following may or may not matter here (measure it using `@btime` ?), but: instead of writing `prod([x for y in z])` you can also write `prod(x for y in z)`. The latter avoids allocating a vector and storing all intermediate results. Whether this is faster or not depends on various factors, but it can be quite substantially if the length (here: `length(t)`) is big.